### DISCUSSION TITLE
ErrorException: Undefined array key "id" - 🚨 out of date assets

### DISCUSSION THREAD

**Discussion by joshhanley** [livewire/livewire#3769] [4. General]
If anyone experiences this error message (in screenshot below), it means **your javascript assets are out of date.**

> Error Exception
> Undefined array key "id" in line 27

![image](https://user-images.githubusercontent.com/882837/133577585-7f87df2e-9412-4ef2-ace8-6d44ff3e5dc2.png)

### To Fix

🚨 **It's only recommended to publish Livewire's javascript assets if you really need to.** If you are unsure, then you probably don't need to.

✅  If you have already published them and **want to un-publish them**, go to this directory `public/vendor/livewire` and delete that directory and everything in it (it should only have `livewire.js`, `livewire.js.map`, and `manifest.json`). 

Livewire does not need the assets to be published to work correctly, so you can safely remove them and Livewire will internally handle the assets for you instead.

⚠️ If you must publish them for whatever reason, you will need to republish them to fix this error, as it is caused by out of date assets (and could happen again in the future).

To do this run `php artisan livewire:publish --assets`.

You should also setup a composer script to do this automatically for you everytime it updates to make sure you always have the latest version.

To do this, add `"@php artisan vendor:publish --force --tag=livewire:assets --ansi"` to your `composer.json` file under `scripts` and `post-autoload-dump`.

See the docs here for more details https://laravel-livewire.com/docs/2.x/installation#publish-assets 

Hope this helps!

**Comment by gavinhewitt** [comment]
But what to do if you have published the assets (which fixes the issue), and 15 mins later they are out of date again? The question is what could be causing them to get outdated.

**Reply by joshhanley** [reply]
@gavinhewitt hmm not sure, that wouldn't make sense as it should use the latest published version. Is there a reason you are publishing them? Can you remove the published version instead?

**Reply by gavinhewitt** [reply]
Gonna be honest, I didn't understand why you wanted me to remove the published assets because I thought that was part of the installation. Went back to the docs and found that publishing assets is optional (and that's what you also explain above):

```
If you prefer the JavaScript assets to be served by your web server not through Laravel, use the livewire:publish command
```
Bingo! I removed the assets and that does the trick indeed. It's still weird why the assets are getting outdated, but at least I don't have to think about that anymore :-)

Thanks!

**Reply by joshhanley** [reply]
@gavinhewitt ah right, that makes sense!

Hmm maybe that's something we should clarify in the docs, as I'd actually recommend not publishing the assets.

Great to hear it worked! Yeah exactly, the less to worry about, the better 🙂

**Reply by joshhanley** [reply]
Have added this to the above 🙂

> Livewire does not need the assets to be published to work correctly, so you can safely remove them and Livewire will internally handle the assets for you instead.

**Reply by dlopez079** [reply]
I appreciate this post as I was stuck with the same error message.  When first reading the documentation, while Laravel mentioned that it was optional, I didn't fully understand the two choices.   I elected to publish but didn't see any gain after publishing.  I un-published and the application is working again.  

Thank you.

**Reply by joshhanley** [reply]
@dlopez079 no worries, glad it helped! 🙂

**Comment by mihaliak** [comment]
Not sure why but the fix is not working for me. I added publish command to composer and error is still happening.

I also removed whole /public/vendor directory but its still happening. Any ideas?

**Reply by joshhanley** [reply]
@mihaliak hmm that seems odd. You don't have something else caching assets like Cloudflare or something like that? Did you try deleting the published assets and seeing if that worked?

**Reply by mihaliak** [reply]
There was one opened browser with old js assets, now working okay. Sorry :)

**Reply by joshhanley** [reply]
No worries! 🙂

**Comment by tiagofrancafernandes** [comment]
Thank you very much

**Comment by bzarzuela** [comment]
This error happens to me even if I didn't publish the Livewire assets

**Reply by bzarzuela** [reply]
I have a theory at #4851

**Comment by flamisz** [comment]
I got the same error, this is what it was.

I had a parent component with a foreach loop. Inside that foreach loop, I had normal html plus calling a child component. I had the `:key` on the child component but I needed to add the `wire:key` to the `foreach` loop too:

```php
@foreach ($events as $event)
  <div wire:key="{{ $event->id }}">

  // ...

  <livewire:promoters.stats :event="$event" :key="$event->id">
  
  // ...
  
  </div>
@endforeach

```
